File: LeetCode/977. Squares of a Sorted Array.java
class Solution {
    public int[] sortedSquares(int[] A) {
        for ( int i = 0; i < A.length; i++ ) {
            A[i] = A[i] * A[i];
        }
        Arrays.sort(A);
        
        return A;
    }
}

class Solution {
    public int[] sortedSquares(int[] A) {
        PriorityQueue<Integer> queue = new PriorityQueue<Integer>();
        
        for ( int i = 0; i < A.length; i++ ) {
                queue.offer(A[i] * A[i]);
        }
        
        for ( int i = 0; i < A.length; i++ ) {
            A[i] = queue.poll();
        }
        
        return A;
    }
}

/* Reference: https://leetcode.com/problems/squares-of-a-sorted-array/discuss/221922/Java-two-pointers-O(N) */

class Solution {
    public int[] sortedSquares(int[] A) {
        int n = A.length;
        int[] result = new int[n];
        int i = 0, j = n - 1;
        for (int p = n - 1; p >= 0; p--) {
            if (Math.abs(A[i]) > Math.abs(A[j])) {
                result[p] = A[i] * A[i];
                i++;
            } else {
                result[p] = A[j] * A[j];
                j--;
            }
        }
        return result;
    }
}

File: LeetCode/977. Squares of a Sorted Array.scala
object Solution {
    def sortedSquares(A: Array[Int]): Array[Int] = {
        var nonNegIndex = 0
        while(nonNegIndex < A.length && A(nonNegIndex) < 0) nonNegIndex += 1
        var i = nonNegIndex
        var j = nonNegIndex - 1
        
        Array.tabulate(A.length) { _ =>
            if(j < 0 || i < A.length && A(i) < -A(j)) {
                val result = A(i) * A(i)
                i += 1
                result
            } else {
                val result = A(j) * A(j)
                j -= 1
                result
            }
        }
    }
}

//https://leetcode.com/problems/squares-of-a-sorted-array/discuss/227197/Elegant-Scala-Solution